Persist the flow destination and use it to redirect users after checkout

niranjan-uma-shankar opened the pull request:
### Changes proposed in this Pull Request

For context, check paObgF-eG-p2

* The post checkout [redirect handler](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/checkout/checkout/index.jsx#L352) does not currently respect the [flow destination](https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/flows-pure.js#L135) that is defined in `flows-pure.js`.
In this PR, we persist the flow destination and use that value to redirect the user after Checkout completes. The value is persisted in a cookie.

Couple of other bugs that this fixes:

* For the same flow, the user ends up in a different destination if they chose a Free plan versus a Paid plan. 
* We ran a `isEligibleForDotComChecklist` check for every destination route, even when the destination was not a checklist.

#### Why cookie and not Redux?

Redux state is not guaranteed to persist across page redirects, as happens when we redirect from the Plan step in signup to the Checkout page, or from WPCOM to the payment handlers such as PayPal or Stripe (thanks @sirreal for the tip).

### Notes:

* There is a separate redirect logic present in the concierge and gsuite upsell nudges([#](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/checkout/concierge-quickstart-session/index.jsx#L346), [#](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/checkout/concierge-session-nudge/index.jsx#L318) and [#](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/checkout/gsuite-nudge/index.jsx#L46)). Removing that logic is a bit more complicated, which is being worked on in https://github.com/Automattic/wp-calypso/pull/33996
UPDATE 07/15: the referenced PRs have been merged. However GSuite upsell nudge is yet to be addressed, which will be taken up as a follow up. So, as of now the Gsuite upsell nudge will redirect without respecting signup destination.
UPDATE 08/16: GSuite is fixed in https://github.com/Automattic/wp-calypso/pull/35384
* There is a known bug with redirect payment types. If a user purchases a plan with a redirect payment type such as PayPal or Sofort, then they will be taken to a Thank You page instead of signup destination. Check [this comment](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-508008978) for context. Bug report for this issue is at #34333 and #34434, and the fix for them will be taken up right after the current PR ships.

**Related:**

An alternate implementation using query params is used in #29136. 

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Scenario 1**

1. Start with the signup flow at /start. 
2. At the site type step, select any type except `Online Store`
3. Without adding any item to cart, select a Free plan and complete signup.
4. Verify that you are redirected to the ~Preview page /view/SITE_SLUG~ checklist page /checklist/SITE_SLUG (see https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-513845666)

Following this, while remaining logged in to your newly created site, do the following:

1.  Click the Add New Site button and enter the site creation flow
2. Select `Online Store` as the site type
3. Add a business or ecommerce plan to cart, and complete checkout.
4. Verify that you are redirected to the website at https://your-website-name.com if on the business plan, or thank you page if on the ecommerce plan.

**Scenario 2**

1. Start at the domain only flow - /start/domain
2. Complete signup and verify that you are taken to the Thank You page

**Scenario 3**

Go through the /start/personal and /start/premium flows. You will see a concierge upsell offer post checkout. Try both scenarios of clicking Yes and No on the upsell nudge. The final destination should be ~site preview~ the checklist page (see https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-513845666)

Go through the /start/business flow. Complete purchase and verify that the final destination is ~site preview~ the checklist.

**Scenario 4**

Signup on Crowdsignal and verify that you are taken to the home page - [here is the Calypso link](http://calypso.localhost:3000/start/crowdsignal/oauth2-name?ref=oauth2&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3Ded297b99225beb5943e02d94cecc9ab3b7d991dfd7a5aed1c52a7333e41aeaf6%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dsignup%2526action%253Drequest_access_token%26wpcom_connect%3D1%26jetpack-code%26jetpack-user-id%3D0%26action%3Doauth2-login&oauth2_client_id=978)

**Scenario 5**

1. Spin up a Jetpack connected instance at https://jurassic.ninja
2. Go through the setup flow (note that you will get redirect to a wordpress.com URL when you click the [`Set up Jetpack` button](https://cloudup.com/cORnykVjiJ8) in wp-admin, so ensure that you stay on calypso.localhost URL).
3. Verify that after purchasing a Jetpack plan, you are taken to  my-plans page `/plans/my-plan/SITE_SLUG?thank-you`

niranjan-uma-shankar committed:
First commit: Store the signup destination in a cookie, and then use that cookie's value to redirect after checkout completes

niranjan-uma-shankar committed:
Add the getSignupDestination function that fetches the destination as either checklist or calypso preview; undo changes that removed a few redux actions from main.jsx;

niranjan-uma-shankar committed:
Adding a few missing prop types

niranjan-uma-shankar committed:
Use the flow destination for gsuite redirect

niranjan-uma-shankar committed:
Add the flow destinatiom for the domain only flow

jonathansadowski reviewed (commented):
> Following this, while remaining logged in to your newly created site, do the following:
> 
> 1. Click the Add New Site button and enter the site creation flow
> 1. Select Online Store as the site type
> 1. Add a business or ecommerce plan to cart, and complete checkout.
> 1. Verify that you are redirected to the website at https://your-website-name.com

This part of the test plan did work for the business plan, but I believe the ecommerce flow circumvents the redirect entirely, and just shows the thank you page (whose action links to woocommerce setup). This is intentional.

ramonjd commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -447,9 +449,10 @@ export class Checkout extends React.Component {
-			return `/${ destination }/${ selectedSiteSlug }${ queryParam }`;
+			const destination = retrieveSignupDestination();
+
+			return `${ destination }${ queryParam }`;
```
Do we need a fallback for the (unlikely) case that the cookie value isn't saved?

niranjan-uma-shankar replied:
Fallback added in [cbf8475](https://github.com/Automattic/wp-calypso/pull/34156/commits/cbf8475fec453f197d97d35f2b5f3bc5bb3943f4).

> Also wondering if it's too defensive to remove the cookie somewhere after we've read it or before we set it? Unless we need it again for future use.

I was oscillating on this point too, if I should clear it or not. I couldn't see a harm in keeping the cookie value, but maybe it's just safer to throw that value away when we don't need it? I'll bite it, and add that fix :)

niranjan-uma-shankar replied:
Fixed in [0f2db23](https://github.com/Automattic/wp-calypso/pull/34156/commits/0f2db23172305e129fdf2348b4526890b67b2144).

ramonjd replied:
> maybe it's just safer to throw that value away when we don't need it? I'll bite it, and add that fix :)

Sorry, I was just musing out loud. Publicly. 😄 

Your judgement on this is way more sophisticated than mine, and I didn't want to give them impression that it was a blocker. 

But since, it doesn't hurt to delete it, bite away 🍭

michaeldcain commented:
@simison: I just wanted to make sure that this is on your radar

niranjan-uma-shankar commented:
Moving this to `In Progress` as I make some changes that I missed.

niranjan-uma-shankar committed:
Add a Thank You destination handler for the ecommerce and ecommerce-onboarding flows

niranjan-uma-shankar committed:
1. Add a default destination if the value from cookie is missing for some reason 2. Rename isElgibleForCheckoutToChecklist to isEligibleForSignupDestination because the destination is no longer just Checklist, so that name does not make sense.

niranjan-uma-shankar committed:
Remove the page builder since we no longer use it

niranjan-uma-shankar committed:
Persist signup destination for the jetpack connect flow

niranjan-uma-shankar committed:
persist the modified destination

niranjan-uma-shankar commented on client/state/selectors/is-eligible-for-checkout-to-checklist.js:
```diff
@@ -10,26 +10,31 @@ import {
 		hasDomainRegistration( cart ) ||
 		hasTransferProduct( cart ) ||
-		( ! hasPlan( cart ) && ! hasConciergeSession( cart ) ) ||
-		hasEcommercePlan( cart )
```
This function will evaluate eligibility even for the eCommerce plan. An eligible plan will be redirected to the flow destination (thank you page for the eCommerce and ecommerce-onboarding flows).

niranjan-uma-shankar commented on client/state/selectors/is-eligible-for-checkout-to-checklist.js:
```diff
@@ -10,26 +10,31 @@ import {
-	return isNewSite( state, siteId ) && isEligibleForDotcomChecklist( state, siteId );
+	const destination = retrieveSignupDestination();
+	if ( destination.includes( '/checklist/' ) ) {
+		return isNewSite( state, siteId ) && isEligibleForDotcomChecklist( state, siteId );
```
We were earlier running the `isEligibleForDotcomChecklist ` check even when the destination when was not checklist. It makes sense to restrict this check only for /checklist destination routes.

niranjan-uma-shankar commented on client/state/selectors/is-eligible-for-checkout-to-checklist.js:
```diff
@@ -10,26 +10,31 @@ import {
  * @return {Boolean} True if current user is able to see the checklist after checkout
  */
-export default function isEligibleForCheckoutToChecklist( state, siteId, cart ) {
+export default function isEligibleForSignupDestination( state, siteId, cart ) {
```
The destination is no longer just Checklist, so it makes sense to rename the function to avoid confusion,

niranjan-uma-shankar commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -387,13 +393,17 @@ export class Checkout extends React.Component {
 			return '/checkout/thank-you/features';
 		}
 
+		if ( this.props.isJetpackNotAtomic ) {
```
Jetpack plan purchases are not shown the Gsuite and Concierge upsell nudges, so it makes sense to move this here at the top of the function.

niranjan-uma-shankar committed:
Changing the destination

niranjan-uma-shankar commented:
> This part of the test plan did work for the business plan, but I believe the ecommerce flow circumvents the redirect entirely, and just shows the thank you page (whose action links to woocommerce setup). This is intentional.

Thanks for pointing that out @jonathansadowski . I've fixed this by adding a thank you destination for the ecommerce flow in the flows definition.

niranjan-uma-shankar commented:
> Moving this to In Progress as I make some changes that I missed.

Update: changes completed, and marking it ready to review again. I've made some changes to the jetpack connect flow too, so appreciate @Automattic/luna 's 👀

niranjan-uma-shankar committed:
1. Rename variable 2. clear the cookie before redirecting

niranjan-uma-shankar committed:
Redirect to signup destination for redirect payment types such as Stripe, which only have a receipt ID

ockham commented:
> ### Testing instructions
> **Scenario 1**
> 
> [...]
> 
> Following this, while remaining logged in to your newly created site, do the following:
> 
> 1. Click the Add New Site button and enter the site creation flow
> 2. Select `Online Store` as the site type
> 3. Add a business or ecommerce plan to cart, and complete checkout.
> 4. Verify that you are redirected to the website at https://your-website-name.com

Selecting the business plan on Step 3, I'm redirected to `/checkout/thank-you/:site/:receiptId`.

I'm not sure that's intended? I read https://github.com/Automattic/wp-calypso/pull/34156#discussion_r297708279 but it seemed like that was meant to apply only to the ecommerce plan. (This comment https://github.com/Automattic/wp-calypso/pull/34156#pullrequestreview-253613466 OTOH refers to the ecommerce _flow_, i.e. maybe the 'Online Store' flow, regardless of plan?)

Pinging @niranjan-uma-shankar and @jonathansadowski for clarification

ockham commented:
> **Scenario 2**
>
> 1. Start at the domain only flow - /start/domain
> 2. Complete signup and verify that you are taken to the Thank You page

This one was fun. I'm only suggested `.live` domains
![image (3)](https://user-images.githubusercontent.com/96308/60270179-201f3580-98f0-11e9-9781-b094fc1824ab.png). I went with the second suggestion (the hyphenated one, for €12.50). I then briefly saw a notice (`Woohoo! You don't owe us anything!`) before being automatically redirected to `/plans`, (which only has a CTA to prompt me to create a WordPress website).

At no point was I asked to pay for my domain. I should mention that this is with the store sandboxed (per PCYsg-IA-p2), which I did to be able to test the scenarios with a dummy credit card (it worked fine with Scenario 1, though).

ockham commented:
> **Scenario 3**
> 
> Go through the /start/personal, /start/premium and /start/business flows. Verify that you are taken to calypso preview after checkout completes.

Verified with `/start/personal` :heavy_check_mark:

ockham commented:
> **Scenario 4**
> 
> Signup on Crowdsignal and verify that you are taken to the home page - [here is the Calypso link](http://calypso.localhost:3000/start/crowdsignal/oauth2-name?ref=oauth2&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3Ded297b99225beb5943e02d94cecc9ab3b7d991dfd7a5aed1c52a7333e41aeaf6%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dsignup%2526action%253Drequest_access_token%26wpcom_connect%3D1%26jetpack-code%26jetpack-user-id%3D0%26action%3Doauth2-login&oauth2_client_id=978)

I used the log in with WordPress.com button and entered the WP.com credentials for my A8c account, and was asked to grant permission to the CrowdSignal app. Afterwards, I was taken to https://app.crowdsignal.com/start. Is that what I was supposed to do?

jonathansadowski commented:
> I'm not sure that's intended? I read [#34156 (comment)](https://github.com/Automattic/wp-calypso/pull/34156#discussion_r297708279) but it seemed like that was meant to apply only to the ecommerce plan. (This comment [#34156 (review)](https://github.com/Automattic/wp-calypso/pull/34156#pullrequestreview-253613466) OTOH refers to the ecommerce _flow_, i.e. maybe the 'Online Store' flow, regardless of plan?)

Thanks. That's correct @ockham -- it should only apply to the ecommerce plan. I don't think we want that behavior for the entire flow. It was actually working as intended previous to the change that added the thank you destination handler, I was just noting that the test plan was inaccurate.

ockham commented:
> **Scenario 5**
> 
> 1. Go through the signup flow and purchase a plan using PayPal or a Stripe source such as Giropay. If a concierge upsell nudge is shown, purchase that too using PayPal or Stripe source.
> 2. Verify that after payment completes, you are redirected to the signup destination

I started at `/start` and purchased a Premium plan during the signup flow. I used a dummy CC and set the country to Turkey, which accordingly to the FG should trigger payment through Stripe. I didn't see a concierge upsel nudge. I ended up at the site preview (`/view/:site`). Is this the correct behavior?

ockham commented:
e2e tests are currently failing, and some of the errors could be conceivably due to the changes in this PR (and not just false positives :sweat_smile: )

niranjan-uma-shankar commented:
bOn the domain only flow - /start/domain:

> This one was fun. I'm only suggested .live domains

This is expected and correct when sandboxed, since the domain suggestions endpoint will only return `.live` domains when sandboxed.

> I then briefly saw a notice (Woohoo! You don't owe us anything!) before being automatically redirected to /plans, (which only has a CTA to prompt me to create a WordPress website).
> At no point was I asked to pay for my domain. 

Could you please share some details on the initial conditions? Were you already logged in  when you ran through this flow? Or was it a fresh start? 

I'm unable to reproduce this though - starting at `/start/domain`, when I click `Just buy a domain` button, I am taken to the Thank You page.  Here's a screen recording - https://cloudup.com/cItgy8kRDcn.

niranjan-uma-shankar commented:
>  started at /start and purchased a Premium plan during the signup flow. I used a dummy CC and set the country to Turkey, which accordingly to the FG should trigger payment through Stripe. I didn't see a concierge upsel nudge. I ended up at the site preview (/view/:site). Is this the correct behavior?

In Scenario 5, I specifically mentioned PayPal and Giropay because they are redirect payment types i.e they take you out of WPCOM to the payment vendor's website. Post payment, they should redirect back to the signup destination. Please check the "Stripe sources with redirect" section in FG - any of those options or PayPal should work to test this scenario. CC does not redirect out of WPCOM so not applicable for this particular scenario. Testing PayPal requires a PayPal sandbox to be configured (instructions in FG). Stripe redirect types such as Giropay or Sofort are easier to test, since with the store sandbox enabled, selecting those payment options automatically take you to the vendor's sandbox requiring no setup from our side. 

The concierge upsell nudge is in an abtest, so you might need to put yourself in the `offer` variation of the `conciergeUpsellDial` test to see it. However, if you choose a redirect payment type for the plan purchase, then the upsell nudge will not be shown. This is a known hack that has been in place since #31102. I've been looking into this, though it may not be part of this PR.

niranjan-uma-shankar committed:
Adding a null check since the destination value could be empty when coming via a free plan

niranjan-uma-shankar committed:
Reverting a change to the ecommerce flow, since we want to show the thank you page for ecommerce plans irrespective of the flow

niranjan-uma-shankar commented:
> Selecting the business plan on Step 3, I'm redirected to /checkout/thank-you/:site/:receiptId.

Thanks for pointing out @ockham and @jonathansadowski  . I've reverted this change in [5b00d92](https://github.com/Automattic/wp-calypso/pull/34156/commits/5b00d92feadce4467d154f9fb290fc3203a7e8a4).

niranjan-uma-shankar committed:
Selecting a free plan now respects the flow destination, so it will show the site preview

niranjan-uma-shankar committed:
The destination for a free plan respects flow destination, whic is the site preview

niranjan-uma-shankar commented:
> e2e tests are currently failing, and some of the errors could be conceivably due to the changes in this PR (and not just false positives 😅 )

All tests pass now.

ockham commented:
> On the domain only flow - /start/domain:
> 
> > This one was fun. I'm only suggested .live domains
> 
> This is expected and correct when sandboxed, since the domain suggestions endpoint will only return `.live` domains when sandboxed.

Thanks for confirming :+1:

> > I then briefly saw a notice (Woohoo! You don't owe us anything!) before being automatically redirected to /plans, (which only has a CTA to prompt me to create a WordPress website).
> > At no point was I asked to pay for my domain.
> 
> Could you please share some details on the initial conditions? Were you already logged in when you ran through this flow? Or was it a fresh start?

No, this was in a newly opened incognito window, i.e. I wasn't logged in.
 
> I'm unable to reproduce this though - starting at `/start/domain`, when I click `Just buy a domain` button, I am taken to the Thank You page. Here's a screen recording - https://cloudup.com/cItgy8kRDcn.

I just tried repro'ing and couldn't :thinking: (But something's still kinda off, the currency I was shown Swedish kroner as the currency, and I have no idea why, since I'm neither located in Sweden, nor do I have any system settings that would hint at Sweden being my locale.)

So I guess it might be safe to ignore, as it's probably not a regression. We should probably keep an eye on this though since maybe there's some cross-request spillage on the server side :thinking:

ockham commented:
> > started at /start and purchased a Premium plan during the signup flow. I used a dummy CC and set the country to Turkey, which accordingly to the FG should trigger payment through Stripe. I didn't see a concierge upsel nudge. I ended up at the site preview (/view/:site). Is this the correct behavior?
> 
> In Scenario 5, I specifically mentioned PayPal and Giropay because they are redirect payment types i.e they take you out of WPCOM to the payment vendor's website. Post payment, they should redirect back to the signup destination. Please check the "Stripe sources with redirect" section in FG - any of those options or PayPal should work to test this scenario. CC does not redirect out of WPCOM so not applicable for this particular scenario. Testing PayPal requires a PayPal sandbox to be configured (instructions in FG). Stripe redirect types such as Giropay or Sofort are easier to test, since with the store sandbox enabled, selecting those payment options automatically take you to the vendor's sandbox requiring no setup from our side.

Gotcha, thanks for clarifying! Proxying through Germany allowed me to select Sofort, which indeed redirected me to a Stripe test payment page :+1: 

However, I was redirected back to `http://calypso.localhost:3000/view/:site/:receiptId` (Or at least I think the last route fragment is the `receiptId` -- it's a 6-digit integer anyway. And that page looks kinda blank (presumably because `/view` routes don't really support the extra route fragment?):

![image](https://user-images.githubusercontent.com/96308/60483865-14998900-9cb9-11e9-8009-e7f6c4a34dbb.png)

niranjan-uma-shankar commented:
> However, I was redirected back to http://calypso.localhost:3000/view/:site/:receiptId (or at least I think the last route fragment is the receiptId -- it's a 6-digit integer anyway). And that page looks kinda blank (presumably because /view routes don't really support the extra route fragment?):

Thanks for catching that @ockham . I investigated more on this and a proper fix depends on fixing an issue we have with redirect payment types. I've P2'ed on p8hgLy-1zx-p2. The guidance I receive on that P2 should also fix the concierge upsell issue I mentioned in [this comment](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-506697266) introduced by the hack in #31102.

niranjan-uma-shankar committed:
Reverting this change as part of removing thank you destination for ecommerce flow

simison commented:
Would be good to see this rebased since it's rebased against 5 days old `master` now.

niranjan-uma-shankar committed:
Reverting a change wherein redirect payment types were made to respect destination; fixing this fully would add complexity to this PR, so it will be taken up as a follow up once this ships

ockham commented:
> > However, I was redirected back to http://calypso.localhost:3000/view/:site/:receiptId (or at least I think the last route fragment is the receiptId -- it's a 6-digit integer anyway). And that page looks kinda blank (presumably because /view routes don't really support the extra route fragment?):
> 
> Thanks for catching that @ockham . I investigated more on this and a proper fix depends on fixing an issue we have with redirect payment types. I've P2'ed on [p8hgLy-1zx-p2](https://href.li/?https://wp.me/p8hgLy-1zx). The guidance I receive on that P2 should also fix the concierge upsell issue I mentioned in [this comment](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-506697266) introduced by the hack in #31102.

Scenario 5 instructions have changed since I last tested, they read quite differently now (involving a Jetpack site now, no more mention of redirecting payment types). Wondering what happened there and what to test now?

niranjan-uma-shankar commented:
>  I investigated more on this and a proper fix depends on fixing an issue we have with redirect payment types. I've P2'ed on p8hgLy-1zx-p2. The guidance I receive on that P2 should also fix the concierge upsell issue I mentioned in this comment introduced by the hack in #31102.

@ockham After looking into this issue, I'm choosing to not include a fix for this issue in the current PR. The reason being that it adds to complexity - it makes this PR bigger and hence more difficult to review. It also requires a backend fix, and hence Payments team to be involved in the review. I prefer noting it as a known bug(updated the PR description with bug report under _Notes_) and take it as a follow up after this PR ships.

Just to reiterate, the known bug is that if a redirect payment type such as PayPal/Sofort/Ebanx is used, then the signup destination might be different from the flow destination sometimes. To clarify, this is the same behaviour in production, and is not a bug introduced by this PR. 

I've updated the test instructions to remove redirect payment types, and replaced Scenario 5 with a Jetpack plan test.

niranjan-uma-shankar commented:
> Would be good to see this rebased again since it's rebased against 5 days old master now.

@simison Rebase done.

niranjan-uma-shankar committed:
Rename file

p-jackson commented on client/signup/main.jsx:
```diff
@@ -224,6 +228,12 @@ class Signup extends React.Component {
 	handleSignupFlowControllerCompletion = ( dependencies, destination ) => {
 		const filteredDestination = getDestination( destination, dependencies, this.props.flowName );
+
+		// If the filtered destination is different from the flow destination, then save it in the store
```
```suggestion
		// If the filtered destination is different from the flow destination (e.g. changes to checkout), then save the flow destination so the user ultimately arrives there
```

Would have saved me a few clicks trying to figure out what `getDestination` does. Alternatively we could rename `getDestination` to something else?

p-jackson commented on client/signup/utils.js:
```diff
@@ -215,3 +216,18 @@ export function canResumeFlow( flowName, progress ) {
 }
+
+export const persistSignupDestination = url => {
+	const options = { path: '/' };
```
I think we should also specify these options too:
- `expires`, because currently this is a session cookie which the browser throws away when it's closed. It feels like this should be persisted like the redux store.
- `sameSite: true`, can't think of a reason why we'd want another site snooping on this cookie value, so we should prevent it.

niranjan-uma-shankar replied:
Thanks for the suggestion, fixed in [4a1de1b](https://github.com/Automattic/wp-calypso/pull/34156/commits/4a1de1b704bc491522cee3315e37f18f8a31508f).

p-jackson commented on client/state/selectors/is-eligible-for-signup-destination.js:
```diff
@@ -18,14 +18,16 @@ import {
  * @param {Object} state Global state tree
  * @param {Number} siteId Site ID
  * @param {Object} cart object
+ * @param {String} destination The final destination URL after signup completes
```
This `destination` param is no longer needed I think.

niranjan-uma-shankar replied:
Fixed in [0a1153f](https://github.com/Automattic/wp-calypso/pull/34156/commits/0a1153f8189218e80b4af70c2218422eda38f6f9).

p-jackson commented on client/signup/config/flows.js:
```diff
@@ -50,14 +50,21 @@ function getRedirectDestination( dependencies ) {
 
-function getChecklistDestination( dependencies ) {
-	return '/checklist/' + dependencies.siteSlug;
+function getSignupDestination( dependencies ) {
```
I think this is a good rename for now, but @Automattic/zelda we might want to consider renaming this back to `getChecklistDestination` when #33448 lands.

p-jackson replied:
I'll resolve this because it's just a team thing.

p-jackson reviewed (changes requested):
Moving on to some manual testing now. I like this change! Much better to have the flows choose the destination than trying to reproduce that ever-changing logic in `getCheckoutCompleteRedirectPath` 👍

p-jackson commented:
Just for context: on `master` the `/start` flow takes you to `/view/{slug}` if the go through the checkout, and `/checklist/{slug}` if you don't.

I didn't realise this was the case! I had just assumed you'd always see the checklist.

This PR makes it consistent in that you always go to `/view/{slug}`, but I'm not sure that's the right option. During user testing @fditrapani has been taking folks through the checklist so I assume that's actually where we want to be taking people, regardless of whether they went through checkout or not.

niranjan-uma-shankar commented:
> Just for context: on master the /start flow takes you to /view/{slug} if the go through the checkout, and /checklist/{slug} if you don't.
I didn't realise this was the case! I had just assumed you'd always see the checklist.

Yes indeed, free and paid plans take to different destinations(as noted in the PR description too), and anyone with a paid plan was never seeing the checklist. Given that a lot of work has been going into improving the checklist experience, this might be a serious bug(unless the destination was intentionally kept that way for some reason). 

> This PR makes it consistent in that you always go to /view/{slug}, but I'm not sure that's the right option. During user testing @fditrapani has been taking folks through the checklist so I assume that's actually where we want to be taking people, regardless of whether they went through checkout or not.

For the moment, I just kept it consistent with how production operates. Please let me know the intended destination(cc @fditrapani ) and I can easily make that switch. Also, please note that we take the user to a "Thank You" page if they have purchased an eCommerce plan.

niranjan-uma-shankar committed:
Update code comment.

Co-Authored-By: Philip Jackson <p-jackson@live.com>

andrewserong commented:
Thank you for the detailed testing steps @niranjan-uma-shankar!

The only issue I ran into testing manually was that after purchasing a domain-only transaction, I was stuck on the loading state of the thank you page, which had an error in the console (screenshot below). Reloading the page resolved the issue, though:
![image](https://user-images.githubusercontent.com/14988353/61264902-80a7e100-a7d1-11e9-9730-0289014e901e.png)

All of the other flows you mentioned worked as expected. I had the same thoughts as @p-jackson regarding redirecting users to the checklist instead of the `/view` page, so good to double-check.

niranjan-uma-shankar commented:
> The only issue I ran into testing manually was that after purchasing a domain-only transaction, I was stuck on the loading state of the thank you page, which had an error in the console (screenshot below). Reloading the page resolved the issue, though:

This is a bug with the domain only flow and not introduced by the PR. I'd reported this issue in p2MSmN-7gP-p2.

andrewserong commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -468,14 +482,8 @@ export class Checkout extends React.Component {
-
-		if ( this.props.isJetpackNotAtomic ) {
-			return `/plans/my-plan/${ selectedSiteSlug }?thank-you&install=all`;
+		if ( this.props.isEligibleForSignupDestination & ( ':receiptId' !== receiptId ) ) {
```
Just a tiny nit here, but does this one need to be a bitwise `&`, or can it be changed to a logical and `&&`?

niranjan-uma-shankar replied:
Thanks for the catch! It was a typo, fixed in [0a1153f](https://github.com/Automattic/wp-calypso/pull/34156/commits/0a1153f8189218e80b4af70c2218422eda38f6f9).

andrewserong commented:
> This is a bug with the domain only flow and not introduced by this PR. I'd reported this issue in p2MSmN-7gP-p2.

Excellent, thank you @niranjan-uma-shankar, and great work on the PR, it's looking good!

niranjan-uma-shankar committed:
Adding the install=all query param for JP checkout

niranjan-uma-shankar committed:
1. Logical AND missing ampersand 2. Remove line from jsdoc

niranjan-uma-shankar committed:
Set the expires and samesite cookie options

niranjan-uma-shankar committed:
Fixes a bug wherein the flow destination was not respected if an upsell nudge is shown after plan checkout

fditrapani commented:
> This PR makes it consistent in that you always go to /view/{slug}, but I'm not sure that's the right option. During user testing @fditrapani has been taking folks through the checklist so I assume that's actually where we want to be taking people, regardless of whether they went through checkout or not.

Ideally, we can specify the destination of the flow in the same place we define flows and not hardcode any particular destination anywhere. 

I'll defer to @michaeldcain on this matter though.

p-jackson reviewed (approved):
LGTM 👍 TIL that cool trick of setting the expiry date in the past to delete a cookie 😎 

Just waiting for @michaeldcain to weigh in on whether users should ultimately be redirected to `/checklist` or `/view` in our "standard" flows. The fact that this logic had diverged is further proof for why this PR is necessary so thank you for doing this!

michaeldcain commented:
@niranjan-uma-shankar @fditrapani: Given that some flows don't result in site creation, I think I'd prefer that we redirect a user to wordpress.com as the fallback. 

Currently this means the reader for most users, but there is discussion/work by @davemart-in's team to make the default destination specific to the user's type and previous behavior.

niranjan-uma-shankar committed:
Fixing merge conflict

davemart-in commented:
> Currently this means the reader for most users, but there is discussion/work by @davemart-in's team to make the default destination specific to the user's type and previous behavior.

A few clarifications just for the record:
- This is just something I have on my list to try and make happen on the side. It's unrelated to my team.
- My proposal is to just do it based on site type. Previous behavior has nothing to do with my proposal.

niranjan-uma-shankar committed:
1. Rebased, fixing a merge conflict 2. Small refactor for better readability

niranjan-uma-shankar committed:
Changing fallback destination to / which is the reader, since not all flows will have a site slug

niranjan-uma-shankar commented:
> Given that some flows don't result in site creation, I think I'd prefer that we redirect a user to wordpress.com as the fallback.

@michaeldcain does this change address your concern? - [11fdade](https://github.com/Automattic/wp-calypso/pull/34156/commits/11fdade6d35b446612ade88c81127efbc0e45949).

Just to re-iterate, once this PR ships, a user going through the onboarding flow will never land on Checklist. Currently on production, only a free plan selection takes to Checklist whereas paid plans take to site preview(except when a eCommerce plan is purchased). This PR will take even free plan users to site preview. Could you confirm this is okay @michaeldcain (cc @shaunandrews).

michaeldcain commented:
> Currently on production, only a free plan selection takes to Checklist whereas paid plans take to site preview

@niranjan-uma-shankar thanks for the clarification. This is not the intended behavior. All users, whether paid or not should end up on the Checklist after Signup completion or Checkout. If that's not easy to do in this PR, @adamziel or @ramonjd can you make sure that change is prioritized?

michaeldcain commented:
I updated that last comment to include "from the onboarding flows"

niranjan-uma-shankar commented:
Thanks for clarifying @michaeldcain . Fortunately, this PR makes it easy to make destination changes :),  so I'll push in that change to make all Onboarding flows go to Checklist. 

Please note behaviour with eCommerce plan.

In production, if a user on the Onboarding flow purchases an eCommerce plan, then we take them a "Processing" page, which then takes to a Thank You page. Here is a recording https://cloudup.com/cSMFYaynO5z.

As @jonathansadowski said in this [comment](https://github.com/Automattic/wp-calypso/pull/34156#pullrequestreview-253613466), the eCommerce plan Thank You page is intended to link to the woocommerce setup, but IMO there are some problems with the approach:

1. Soon after checkout, the "Thank You" page presents a CTA to resend email

<img width="715" alt="Screenshot 2019-07-19 at 10 28 53 PM" src="https://user-images.githubusercontent.com/1269602/61552130-a0aaff00-aa74-11e9-8c7a-99a792b62535.png">

2. If you click the Confirm Email button in your email inbox, you will be taken to `https://wordpress.com/plans/my-plan/{SITE_SLUG}/?verified=1`.

3. To see the woocommerce setup button, the user will have to keep the original Thank You page browser tab open, and navigate to that tab after following the confirm link in their email. The Resend Email button on the Thank You page morphs into a `Create Store` button, but it's useful only if the user remembers to comes here:

<img width="1399" alt="Screenshot 2019-07-19 at 10 26 16 PM" src="https://user-images.githubusercontent.com/1269602/61552246-ea93e500-aa74-11e9-904b-ca035cc0e935.png">

Would you like to retain the current behaviour of taking eCommerce plan purchasers to "Thank You" page, or would you want even eCommerce purchasers to go to Checklist? If the latter, it's a simple fix of removing this [line](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/is-eligible-for-checkout-to-checklist.js#L35), but maybe we need a eCommerce plan specific checklist in that case? If the former, then maybe we should fix the Confirm Email link to point to the thank you page having the `Create Store` CTA instead of taking to /plans page(out of scope for this PR).

jonathansadowski commented:
> If you click the Confirm Email button in your email inbox, you will be taken to `https://wordpress.com/plans/my-plan/{SITE_SLUG}/?verified=1`

You should be taken to `/wp-admin/admin.php?page=wc-setup` on your site upon clicking the confirm email button (see D29483-code), which is the same action as the `Create your store!` button.

> Would you like to retain the current behaviour of taking eCommerce plan purchasers to "Thank You" page, or would you want even eCommerce purchasers to go to Checklist? If the latter, it's a simple fix of removing this line, but maybe we need a eCommerce plan specific checklist in that case? If the former, then maybe we should fix the Confirm Email link to point to the thank you page having the Create Store CTA instead of taking to /plans page(out of scope for this PR).

A ecommerce checklist was discussed in the potential scope of paObgF-iF-p2, but we punted on that. I think, unless we have an ecommerce-specific checklist, we want the current behavior retained, but I'll defer to @michaeldcain on that one.

ramonjd commented:
> All users from the onboarding flows, whether paid or not should end up on the Checklist after Signup completion or Checkout. If that's not easy to do in this PR, @adamziel or @ramonjd can you make sure that change is prioritized?

> so I'll push in that change to make all Onboarding flows go to Checklist.

Sounds like you got it @niranjan-uma-shankar 👍 

Let us know if this requirement needs to be handled in another PR

shaunandrews commented:
Perhaps unrelated to this PR: The purple icon on the green background makes me cry. We don't really use those icons anymore, perhaps we can replace it with something else, or just remove it entirely.

michaeldcain commented:
@niranjan-uma-shankar: you and Jon are correct, let's leave the ecommerce  and ecommerce flow destinations alone for now.

> If the former, then maybe we should fix the Confirm Email link to point to the thank you page having the Create Store CTA instead of taking to /plans page(out of scope for this PR).

I think @Automattic/isotope is looking at a Hack week sprint on the flow - this might fit nicely there. /cc @timmyc

niranjan-uma-shankar committed:
Making Checklist the default destination for the Onboarding flow

niranjan-uma-shankar commented:
Default destination for the Onboarding flow is now Checklist (fixed in [cf9e038](https://github.com/Automattic/wp-calypso/pull/34156/commits/cf9e0387b52f70edf140f980152d45dc487036e4)).

I'll await for a ✅from @Automattic/luna before deploying since this touches Jetpack onboarding. 

> Perhaps unrelated to this PR: The purple icon on the green background makes me cry. We don't really use those icons anymore, perhaps we can replace it with something else, or just remove it entirely.

@shaunandrews If you don't mind, I'll take it up as a follow up PR.

ockham commented:
> ### Testing instructions
> **Scenario 1**
> 
> 1. Start with the signup flow at /start.
> 2. At the site type step, select any type except `Online Store`
> 3. Without adding any item to cart, select a Free plan and complete signup.
> 4. Verify that you are redirected to the ~Preview page /view/SITE_SLUG~ checklist page /checklist/SITE_SLUG (see [#34156 (comment)](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-513845666))

:+1:
 
> Following this, while remaining logged in to your newly created site, do the following:
> 
> 1. Click the Add New Site button and enter the site creation flow
> 2. Select `Online Store` as the site type
> 3. Add a business or ecommerce plan to cart, and complete checkout.
> 4. Verify that you are redirected to the website at https://your-website-name.com if on the business plan, or thank you page if on the ecommerce plan.

:+1:

ockham commented:
> **Scenario 2**
> 
> 1. Start at the domain only flow - /start/domain
> 2. Complete signup and verify that you are taken to the Thank You page

I seem to be stuck at the "Give us a minute and we'll move right along." screen. Here's a possibly related error I'm seeing in my browser console:

![image](https://user-images.githubusercontent.com/96308/61649498-67bc9580-acd3-11e9-96d8-153db5f3a994.png)

niranjan-uma-shankar commented:
> I seem to be stuck at the "Give us a minute and we'll move right along." screen. Here's a possibly related error I'm seeing in my browser console:

Nice catch @ockham . This issue appears for me on production so is unrelated to this PR. I'm going to check in with domains dev nevertheless.

ockham commented:
> **Scenario 3**
> 
> Go through the /start/personal and /start/premium flows. You will see a concierge upsell offer post checkout. Try both scenarios of clicking Yes and No on the upsell nudge. The final destination should be ~site preview~ the checklist page (see https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-513845666)

:+1: (I didn't get the upsell nudge)

> Go through the /start/business flow. Complete purchase and verify that the final destination is ~site preview~ the checklist.

:+1: 

(Hope that this is really also supposed to take me to the checklist)

ockham commented:
> **Scenario 4**
> 
> Signup on Crowdsignal and verify that you are taken to the home page - [here is the Calypso link](http://calypso.localhost:3000/start/crowdsignal/oauth2-name?ref=oauth2&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3Ded297b99225beb5943e02d94cecc9ab3b7d991dfd7a5aed1c52a7333e41aeaf6%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dsignup%2526action%253Drequest_access_token%26wpcom_connect%3D1%26jetpack-code%26jetpack-user-id%3D0%26action%3Doauth2-login&oauth2_client_id=978)

I used the log in with WordPress.com button and entered the WP.com credentials for my A8c account. Afterwards, I was taken to https://app.crowdsignal.com/dashboard. Is that what I was supposed to do? (I'd granted permission to use my account to CrowdSignal during a previous test run.)

ockham commented:
> **Scenario 5**
> 
> 1. Spin up a Jetpack connected instance at https://jurassic.ninja
> 2. Go through the setup flow (note that you will get redirect to a wordpress.com URL when you click the [`Set up Jetpack` button](https://cloudup.com/cORnykVjiJ8) in wp-admin, so ensure that you stay on calypso.localhost URL).
> 3. Verify that after purchasing a Jetpack plan, you are taken to  my-plans page `/plans/my-plan/SITE_SLUG?thank-you`

:+1:

ockham reviewed (approved):
The (apparently pre-existing) [issue with the domain flow](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-513867612) aside, this appears to be working as described :+1:

I've noticed that the e2e test canary is failing again. From a quick glance, it seems like that the failure might reflect the changed flow redirects, and that the tests thus might need updating. Team Calypso might be able to help with that.

This is a big enough change with potential for regressions that we should alert Happiness Engineers before merging so they know who to talk to in case any signup flows break.

Feel free to merge after fixing e2e tests and alerting Happiness Engineers!

niranjan-uma-shankar commented:
> I used the log in with WordPress.com button and entered the WP.com credentials for my A8c account. Afterwards, I was taken to https://app.crowdsignal.com/dashboard. Is that what I was supposed to do? (I'd granted permission to use my account to CrowdSignal during a previous test run.)

To be honest, I'm just going by the current behaviour on production.
1. When using the `Create a new WordPress.com account` option, it takes me to https://app.crowdsignal.com/start?plan=free on both production and this PR.
2. From your old comment [here](https://github.com/Automattic/wp-calypso/pull/34156#issuecomment-506350680), log in with WP took you to /start which aligns with the behaviour on production.
3. Since you are  signing in again with the same credentials, it makes sense that you were taken to the dashboard. However, that may not be a sign up flow. The objective is to just ensure that signup flow destination for Crowdsignal remains unaffected (compared with the behaviour on production).

niranjan-uma-shankar committed:
Revert e2e test changes since the default destination is now checklist

niranjan-uma-shankar committed:
Making checklist the default destination for both the main and onboarding flows

niranjan-uma-shankar committed:
Fixed a minor bug.
If you have some item in the cart(could be anything), and then open a concierge upsell nudge by the direct link to /offer-support-session, then clicking Skip button on the upsell nudge would take you to a Thank You page.
